validator/client/beacon-api/stream_blocks.go

```diff
@@ -0,0 +1,196 @@
+package beacon_api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	"github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+type abstractSignedBlockResponseJson struct {
+	Version             string          `json:"version" enum:"true"`
+	ExecutionOptimistic bool            `json:"execution_optimistic"`
+	Finalized           bool            `json:"finalized"`
+	Data                json.RawMessage `json:"data"`
+}
+
+type streamBlocksAltairClient struct {
+	grpc.ClientStream
+	ctx                 context.Context
+	beaconApiClient     beaconApiValidatorClient
+	streamBlocksRequest *ethpb.StreamBlocksRequest
+	prevBlockSlot       primitives.Slot
+	pingDelay           time.Duration
+}
+
+type headSignedBeaconBlockResult struct {
+	streamBlocksResponse *ethpb.StreamBlocksResponse
+	executionOptimistic  bool
+	slot                 primitives.Slot
+}
+
+func (c beaconApiValidatorClient) streamBlocks(ctx context.Context, in *ethpb.StreamBlocksRequest, pingDelay time.Duration) ethpb.BeaconNodeValidator_StreamBlocksAltairClient {
+	return &streamBlocksAltairClient{
+		ctx:                 ctx,
+		beaconApiClient:     c,
+		streamBlocksRequest: in,
+		pingDelay:           pingDelay,
+	}
+}
+
+func (c *streamBlocksAltairClient) Recv() (*ethpb.StreamBlocksResponse, error) {
+	result, err := c.beaconApiClient.getHeadSignedBeaconBlock(c.ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get latest signed block")
+	}
+
+	// We keep querying the beacon chain for the latest block until we receive a new slot
+	for (c.streamBlocksRequest.VerifiedOnly && result.executionOptimistic) || c.prevBlockSlot == result.slot {
+		select {
+		case <-time.After(c.pingDelay):
+			result, err = c.beaconApiClient.getHeadSignedBeaconBlock(c.ctx)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to get latest signed block")
+			}
+		case <-c.ctx.Done():
+			return nil, errors.New("context canceled")
+		}
+	}
+
+	c.prevBlockSlot = result.slot
+	return result.streamBlocksResponse, nil
```

Comment:
Can you explain the flow on a high level? I am a bit puzzled how these components are tied together:
- `streamBlocksAltairClient`
- `Recv()`
- `result.streamBlocksResponse`

Reply:
It's similar to the approach we took in `activation.go`. `StreamBlocksAltair` returns a `streamBlocksAltairClient` (non-blocking) which implements the `grpc.ClientStream` interface. When the caller calls `Recv()` on it, it does a blocking call and polls the beacon node until a new block is returned. `Recv()` returns an `ethpb.StreamBlocksResponse`.